### PR TITLE
POC -- adopt eslint-plugin-react-compiler

### PR DESCRIPTION
**Pull request by grahamlangford**
## What does this PR do?

- Adds the react-compiler eslint plugin
- Attempts fixes for some of the errors

## Learnings

- One of the biggest issues flagged by the plugin was accessing `ref.current` during renders
- `usePreviousValue` used to be recommended by React, but now it's discouraged
    - See: https://legacy.reactjs.org/docs/hooks-faq.html#what-can-i-do-if-my-effect-dependencies-change-too-often:~:text=We%20have%20previously,predict%20and%20understand. 

## Potential Slices

- Enable the rule on "warn"
- Remove `usePreviousValue`
- Update `useMemoCompare` to pass check
    - Validate sufficient unit test coverage exists 
- Update `useMergeAsyncState` to pass check
    - Validate sufficient unit test coverage exists 
- Update `useOnMountOnly` to pass check
    - Deprecate/flag it as bad practice
- Enable exhaustive-deps where disabled
    - If the concern is a potentially unmemoized callback, can be left disabled until react-compiler is implemented   
- Eliminate any hacks like `objectHash` for deps arrays 
    

For more information on our expectations for the PR process, see the
[code review principles doc](https://www.notion.so/pixiebrix/Code-Review-Principles-1ce7276b82a84d2a995d55ad85e1310d?pvs=4)

### CODE REVIEW

File: src/components/DatabaseUnresponsiveBanner.tsx [hunk 1]
```diff
@@ -47,13 +46,17 @@ const DatabaseUnresponsiveBanner: React.VoidFunctionComponent<{
...
-    reportError(new Error("IDB unresponsive"));
-  }
+  useEffect(() => {
+    if (showBanner && !showBannerRef.current) {
+      reportEvent(Events.IDB_UNRESPONSIVE_BANNER);
+      reportError(new Error("IDB unresponsive"));
+
+      showBannerRef.current = true;
+    }
+  }, [showBanner]);
```
Comment (grahamlangford) [inline src/components/DatabaseUnresponsiveBanner.tsx:59]: Used this pattern to replace `usePreviousValue` in multiple places

File: src/hooks/useMemoCompare.ts
Comment (grahamlangford) [inline src/hooks/useMemoCompare.ts:1]: Fortunately we have good unit test coverage here, so the changes should be safe

File: src/pageEditor/tabs/editTab/dataPanel/tabs/BrickOutputTab.tsx
Comment (grahamlangford) [inline src/pageEditor/tabs/editTab/dataPanel/tabs/BrickOutputTab.tsx:1]: Unrelated: BrickPreview wasn't using starterbrick

File: src/pageEditor/tabs/effect/BrickPreview.tsx [hunk 2]
```diff
@@ -202,8 +199,7 @@ const BrickPreview: React.FunctionComponent<{
     if ((context && brickInfo?.isPure) || brickInfo?.traceOptional) {
       void debouncedRun(brickConfig, context as unknown as BrickArgsContext);
     }
-    // eslint-disable-next-line react-hooks/exhaustive-deps -- using objectHash for context
-  }, [debouncedRun, brickConfig, brickInfo, objectHash(context ?? {})]);
+  }, [debouncedRun, brickConfig, brickInfo, context]);
```
Comment (grahamlangford) [inline src/pageEditor/tabs/effect/BrickPreview.tsx:202]: objectHash isn't necessary if we ensure context is referentially equal

File: src/components/DatabaseUnresponsiveBanner.tsx (hunk 1 above)
Comment (twschiller) [reply src/components/DatabaseUnresponsiveBanner.tsx:59]: I believe this analytics case should just be a `useEffect`: https://react.dev/learn/synchronizing-with-effects#sending-analytics

```
useEffect(() => {
    if (showBanner) {
      reportEvent(Events.IDB_UNRESPONSIVE_BANNER);
      reportError(new Error("IDB unresponsive"));
    }
  }, [showBanner]);
```

**Comment by twschiller** [comment]
@grahamlangford Looks like we can partially role out the compiler / it automatically can skip parts that aren't compatible yet: https://react.dev/learn/react-compiler#using-the-compiler-effectively

**Comment by grahamlangford** [comment]
> @grahamlangford Looks like we can partially role out the compiler / it automatically can skip parts that aren't compatible yet: https://react.dev/learn/react-compiler#using-the-compiler-effectively

Yes, I saw that as well. This draft PR was an experiment to see how difficult it would be to make it compliant. Since it also impacted useEffect usage as well, I figured it might also help with the React 18 upgrade. 

Not planning to spend much more time on this PR. If whatever is breaking the e2e tests is an easy fix, then I'll see about making that change. But if not, I'll close it

File: src/components/DatabaseUnresponsiveBanner.tsx (hunk 1 above)
Comment (grahamlangford) [reply src/components/DatabaseUnresponsiveBanner.tsx:59]: I intended this to replicate the `usePreviousValue` version as much as possible, but I would need an `else { showBannerRef.current = false }` to properly replicate. I think your suggestion is all we actually need though

**Pull request closed without merging**